src/singlelayerqg.jl

```diff
@@ -0,0 +1,508 @@
+  @. vars.vh =  im * grid.kr * vars.ψh
+
+  ldiv!(vars.q, grid.rfftplan, vars.qh)
+  ldiv!(vars.u, grid.rfftplan, vars.uh)
+  ldiv!(vars.v, grid.rfftplan, vars.vh)
+
+  uq_plus_η = vars.u                                            # use vars.u as scratch variable
+  @. uq_plus_η *= vars.q + params.eta                           # u*(q+η)
+  vq_plus_η = vars.v                                            # use vars.v as scratch variable
+  @. vq_plus_η *= vars.q + params.eta                           # v*(q+η)
+
+  uq_plus_ηh = vars.uh                                          # use vars.uh as scratch variable
+  mul!(uq_plus_ηh, grid.rfftplan, uq_plus_η)                    # \hat{u*(q+η)}
+  vq_plus_ηh = vars.vh                                          # use vars.vh as scratch variable
+  mul!(vq_plus_ηh, grid.rfftplan, vq_plus_η)                    # \hat{v*(q+η)}
+
+  @. N = -im * grid.kr * uq_plus_ηh - im * grid.l * vq_plus_ηh  # -∂[u*(q+η)]/∂x -∂[v*(q+η)]/∂y
+  return nothing
+end
+
+function calcN!(N, sol, t, clock, vars, params, grid)
+  calcN_advection!(N, sol, t, clock, vars, params, grid)
+  addforcing!(N, sol, t, clock, vars, params, grid)
+
+  return nothing
+end
+
+addforcing!(N, sol, t, clock, vars::Vars, params, grid) = nothing
+
+function addforcing!(N, sol, t, clock, vars::ForcedVars, params, grid)
+  params.calcF!(vars.Fh, sol, t, clock, vars, params, grid)
+  @. N += vars.Fh
+
+  return nothing
+end
+
+function addforcing!(N, sol, t, clock, vars::StochasticForcedVars, params, grid)
+  if t == clock.t # not a substep
+    @. vars.prevsol = sol # sol at previous time-step is needed to compute budgets for stochastic forcing
+    params.calcF!(vars.Fh, sol, t, clock, vars, params, grid)
+  end
+
+  @. N += vars.Fh
+
+  return nothing
+end
+
+"""
+    streamfunctionfrompv!(ψh, qh, params, grid)
+
+Invert the Fourier transform of PV to obtain the Fourier transform of the streamfunction `ψh`.
+"""
+function streamfunctionfrompv!(ψh, qh, params::BarotropicQGParams, grid)
+  @. ψh =  - grid.invKrsq * qh
+end
+
+function streamfunctionfrompv!(ψh, qh, params::EquivalentBarotropicQGParams, grid)
+  @. ψh  = - qh / (grid.Krsq + 1 / params.deformation_radius^2)
+end
+
+
+# ----------------
+# Helper functions
+# ----------------
+
+"""
+    updatevars!(sol, vars, params, grid)
+
+Update the variables in `vars` with the solution in `sol`.
+"""
+function updatevars!(sol, vars, params, grid)
+  @. vars.qh = sol
+  streamfunctionfrompv!(vars.ψh, vars.qh, params, grid)
+  @. vars.uh = -im * grid.l  * vars.ψh
+  @. vars.vh =  im * grid.kr * vars.ψh
+
+  ldiv!(vars.q, grid.rfftplan, deepcopy(vars.qh))
+  ldiv!(vars.ψ, grid.rfftplan, deepcopy(vars.ψh))
+  ldiv!(vars.u, grid.rfftplan, deepcopy(vars.uh))
+  ldiv!(vars.v, grid.rfftplan, deepcopy(vars.vh))
+
+  return nothing
+end
+
+updatevars!(prob) = updatevars!(prob.sol, prob.vars, prob.params, prob.grid)
+
+"""
+    set_q!(prob, q)
+    set_q!(sol, vars, params, grid)
```

Comment:
I guess it makes sense just to document the user API here via `set_q!(prob, q)` rather than also including the "lower-level" form.

Reply:
Note there's also a typo in the lower level form, should be `set_q!(sol, vars, params, grid, q)`

Reply:
You are suggesting that `set_q!(sol, vars, params, grid, q)` is useless and we just need to have `set_q!(prob, q)`? I think I agree... I don't remember where `set_q!(sol, vars, params, grid, q)` was useful for... perhaps legacy now...